lib/capability/capability.go

```diff
@@ -0,0 +1,94 @@
+package capability
+
+import (
+	"fmt"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const CapabilityAnnotation = "capability.openshift.io/name"
+
+type ClusterCapabilities struct {
+	KnownCapabilities map[configv1.ClusterVersionCapability]struct{}
+
+	// value indicates whether capability was explicitly included
+	EnabledCapabilities map[configv1.ClusterVersionCapability]bool
+}
+
+func SetCapabilities(config *configv1.ClusterVersion) ClusterCapabilities {
+	var capabilities ClusterCapabilities
+	capabilities.KnownCapabilities = setKnownCapabilities(config)
+	capabilities.EnabledCapabilities = setEnabledCapabilities(config)
+	return capabilities
+}
+
+func GetCapabilitiesStatus(capabilities ClusterCapabilities) configv1.ClusterVersionCapabilitiesStatus {
+	var status configv1.ClusterVersionCapabilitiesStatus
+	for k, _ := range capabilities.EnabledCapabilities {
+		status.EnabledCapabilities = append(status.EnabledCapabilities, k)
+	}
+	for k, _ := range capabilities.KnownCapabilities {
+		status.KnownCapabilities = append(status.KnownCapabilities, k)
+	}
+	return status
+}
+
+func CheckResourceEnablement(annotations map[string]string, capabilities ClusterCapabilities) error {
+	val, ok := annotations[CapabilityAnnotation]
+	if ok {
+		caps := strings.Split(val, "+")
+		numCaps := len(caps)
+		unknownCaps := make([]string, numCaps)
+		disabledCaps := make([]string, numCaps)
```

Comment:
I think you want `make([]string, 0, numCaps)` for both of these to say "I want to allocate room for `numCaps` items, but I don't actually have any items for this slice yet".  As it stands, you are initializing `numCaps` empty-string items, and then below you're appending additional caps after those empty-string items.  [Go play][1].

[1]: https://go.dev/play/p/ekvG3pqKMm5